Add comprehensive specs for tool execution failure handling

emk-klaay opened the pull request:
- Add specs to ensure conversation continues when tool calls fail
- Test error logging and error message propagation as tool responses
- Verify state remains :in_progress to allow conversation continuation
- Cover different error types: StandardError, ArgumentError, RuntimeError
- Test edge cases: tool not found, missing tool call IDs, multiple failures
- Ensure assistant can continue processing after tool failures

emk-klaay committed:
Add comprehensive specs for tool execution failure handling

- Add specs to ensure conversation continues when tool calls fail
- Test error logging and error message propagation as tool responses
- Verify state remains :in_progress to allow conversation continuation
- Cover different error types: StandardError, ArgumentError, RuntimeError
- Test edge cases: tool not found, missing tool call IDs, multiple failures
- Ensure assistant can continue processing after tool failures

alb-klaay commented on lib/langchain/assistant.rb:
```diff
@@ -322,7 +322,11 @@ def execute_tools
-      :failed
+      
+      # Add the error message as a tool response so the AI can see what went wrong
+      add_message(role: @llm_adapter.tool_role, content: "Error: #{e.message}", tool_call_id: messages.last.tool_calls.first&.dig("id"))
```
How do we know it's the first tool call that threw an error?

alb-klaay replied:
Maybe we should move the rescue to the `run_tool` method, since we have required info there?

https://github.com/klaayinc/langchainrb/blob/f7de8f6b62a2c525817cac93a702a3b732173e63/lib/langchain/assistant.rb#L372

emk-klaay replied:
Yeah, I also thought about that.

emk-klaay committed:
Refactor tool error handling to individual tool level

- Move error handling from execute_tools to run_tool method
- Each tool failure is now handled independently
- Better error isolation: one tool failure doesn't stop others
- More specific error logging with tool name
- Update specs to reflect new error handling approach
- All tool failure scenarios continue conversation properly

emk-klaay committed:
Improve error handling consistency in rescue block

- Use same if/else pattern in rescue block as success case
- Handle ToolResponse objects consistently in error scenarios
- Preserve image_url when tool returns ToolResponse but fails during processing
- Maintain consistency between success and error handling paths
- All specs continue to pass with improved error handling

emk-klaay committed:
Update error message format and improve error handling consistency

- Change error message format from 'Error: {message}' to 'Error running tool: {message}'
- Use consistent if/else pattern in rescue block matching success case
- Handle ToolResponse objects consistently in error scenarios
- Preserve image_url when tool returns ToolResponse but fails during processing
- Update all specs to match new error message format
- All tool failure scenarios continue conversation properly

emk-klaay commented on lib/langchain/assistant.rb:
```diff
@@ -374,6 +371,17 @@ def run_tool(tool_call)
+      
+      # Add the error message as a tool response so the AI can see what went wrong
+      # Handle both ToolResponse and legacy return values for error cases
+      content = "Error running tool: #{e.message}"
```
```suggestion
      content = "Error running tool (#{tool_name}): #{e.message}"
```

emk-klaay committed:
Apply suggestion from @emk-klaay

emk-klaay committed:
Add Sentry error tracking and improve error message format

- Add Sentry.capture_exception with comprehensive context data
- Include tool_call, assistant state, messages, metadata, and configuration
- Improve error message format to include tool name: 'Error running tool (tool_name): message'
- Maintain consistent error handling for both ToolResponse and legacy return values
- All error context is captured for better debugging and monitoring
